hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/FSSchedulerNode.java

```diff
@@ -131,10 +203,58 @@ void addContainersForPreemption(Collection<RMContainer> containers) {
 
   /**
    * Remove container from the set of containers marked for preemption.
+   * Reserve the preempted resources for the app that requested
+   * the preemption.
    *
    * @param container container to remove
    */
   void removeContainerForPreemption(RMContainer container) {
     containersForPreemption.remove(container);
   }
+
+  /**
+   * The Scheduler has allocated containers on this node to the given
+   * application.
+   * @param rmContainer Allocated container
+   * @param launchedOnNode True if the container has been launched
+   */
+  @Override
+  protected synchronized void allocateContainer(RMContainer rmContainer,
+                                                boolean launchedOnNode) {
+    super.allocateContainer(rmContainer, launchedOnNode);
+    Resource allocated = rmContainer.getAllocatedResource();
+    if (!Resources.isNone(allocated)) {
+      for (FSAppAttempt app: resourcesPreemptedPerApp.keySet()) {
```

Comment:
A for loop seems a little wasteful. Would it make any sense to keep another map so this can just be a lookup?